api: Implement missing/empty fields for MVP

mitjat opened the pull request:
Resolves #256 

Implements all the fields that were added to the API in #247.

Most of the fields were straightforward, except for runtime balances. Changes introduced there:
- Spin-off PR #288 to implement tracking of oasis-sdk balances in runtimes
- Moved `/emerald/tokens` to `/emerald/evm_tokens`; the former should be reserved for oasis-sdk tokens (currently only ROSE); internally renamed `RuntimeToken`->`EvmToken`
- Expanded `/consensus/account` to return the balances across all paratimes, simultaneously for oasis-sdk tokens and EVM tokens.

The last two points are not entirely coherent with each other: one splits the two kinds of tokens across two different URLs, while the other one unifies them. Both approaches have pluses and minuses; in the end, I decided for this combo because I think it best mirrors what the frontend needs, and we can always add endpoints that slice and dice and present (almost) the same information slightly differently.

I recommend reviewing commit by commit, as many of them are boilerplate.

mitjat commented on api/spec/v1.yaml:
```diff
@@ -1341,12 +1341,12 @@ components:
+          format: byte
           description: |
-            The hex-encoded data from which the oasis address was derived.
+            The base64-encoded data from which the oasis address was derived.
```
Yeah. `format: hex` is not an official openapi format. openapi allows you to put any value for `format`, but unless it's one of the few official ones, it's just informative.
`format: byte` is one of the official ones; it means "the contents are a raw byte sequence, and clients should encode/decode it with base64.

This goes against what would be natural for eth addresses. If this field always contained an eth address, I would say screw it and do the hex serialization by hand. But this field can also contain ed25519 pubkeys (and maybe runtime hashes in some cases? not sure). So the natural lowest common denominator is raw bytes.
In my understanding, if the autogenerated TS client is openapi compliant, it will actually parse the base64 automatically and give consumers the raw bytes, which is not too bad even for eth addresses.

Andrew7234 replied:
got it, thanks!

mitjat commented on storage/client/client.go:
```diff
@@ -656,11 +665,17 @@ func (c *StorageClient) Account(ctx context.Context, address staking.Address) (*
+		// when filling them from the database.
+		Allowances:                  []Allowance{},
+		AddressPreimage:             &AddressPreimage{},
+		DelegationsBalance:          &common.BigInt{},
+		DebondingDelegationsBalance: &common.BigInt{},
```
Our docs say: "For efficiency, this field is omitted when listing multiple-accounts."
It's because we need a join and a sum (IIRC) to derive these.

We could have separate types for Account and AccountThatAppearsInAnAccountList, but ... not worth the code duplication IMO.

Andrew7234 replied:
Er but since this is the `Account` endpoint (not `Accounts`), the `AccountQuery` only fetches one account and so it will never be null due to the `COALESCE`. Can we remove these two lines here and simply treat `DelegationsBalance` and `DebondingDelegationsBalance` like the other fields, eg 
```
                a.Debonding, err = c.numericToBigInt(ctx, &debondingNum)
		if err != nil {
			return nil, apiCommon.ErrStorageError
		}
		a.DelegationsBalance, err = c.numericToBigInt(ctx, &delegationsBalanceNum)
		if err != nil {
			return nil, apiCommon.ErrStorageError
		}
		a.DebondingDelegationsBalance, err = c.numericToBigInt(ctx, &debondingDelegationsBalanceNum)
		if err != nil {
			return nil, apiCommon.ErrStorageError
		}
```

mitjat replied:
I see. Unfortunately not, because `numericToBigInt` returns a `BigInt`, so we need to assign it to something of type `BigInt`. 
`a.DelegationsBalance` is `*BigInt` (enforced by the yaml), so we need to assign with a `*` to make the type checker happy:
```
*a.DelegationsBalance, err = c.numericToBigInt(ctx, &delegationsBalanceNum)
``` 
and dereferencing `a.DelegationsBalance` when it's `nil` (which it will be without the line that started this comment thread) leads to having a bad time.

mitjat commented on storage/client/queries.go:
```diff
@@ -46,14 +46,17 @@ func (qf QueryFactory) TransactionsQuery() string {
 				%[1]s.transactions.body as body,
-				%[1]s.transactions.code as code
+				%[1]s.transactions.code as code,
+				%[1]s.blocks.time as time
```
I don't think the MVP mocks have it anywhere, so let's add it when/if needed.

I am actually itching to remove some of the filters we already have. Every filter needs a new index (and they make inserts/updates expensive). Some of the filters we have _don't_ have a filter, so they will result in horribly slow queries that will gunk up our DB. Again, the more filters we have, the more likely a user can create an expensive query by using that filter (or a creative filter combo), either because they're actively trying to be evil, or (more likely) because we didn't prepare sufficiently for that kind of query.

To your point, as a user, I do want a time filter in the GUI. So we might well end up adding it eventually.

Andrew7234 replied:
Agreed on both points; will add to the product meeting agenda.

mitjat reviewed (commented):
Thanks Andy!

Andrew7234 reviewed (approved):
Looks good pending one comment.

Would appreciate if someone more familiar with the token logic could take a second pass through

mitjat commented:
Thank again, Andy.
In the end, I added one more commit/change to the PR: I tweaked the presentation (= API output) of runtime tokens in that I split the formerly single list into two lists: oasis-sdk tokens and EVM tokens. The former merging always felt a little forced (as also alluded to in the PR description). Now I noticed that our UI mocks require yet another EVM-token specific field (`name`, which oasis-sdk tokens don't have) and that pushed me to fully separate the `EvmBalance` and `SdkBalance` types.
